### PR TITLE
Implement assembly version's per "band" to ensure binding redirects work

### LINKED ISSUE #205: Implement assembly version strategy
After a long discussion with @ericstj about #97, it seems like there's no really good options here, just one possibly "least bad" one.

To summarize the issue, when a package provides multiple assemblies that work on a platform the AssemblyVersion for each "higher/wider" version needs to be greater to ensure binding redirects work as expected. This means we cannot simply use the NuGet package version for all assemblies in the package. This is true whether it's `net40` and `net45` or `netstandard1.0`, `netstandard1.1` and `netstandard1.3` as we have today. The core issue is the same -- higher versions have a superset of the functionality as lower-platform versions.

In assembly versioning, it's generally considered the case that a higher assembly version contains all of the functionality of a lower one. This leads to the following "least bad" option:

- We will keep the major version of the assembly equal to the NuGet package (this is the one area we may have to reconsider)
- For each assembly in a compatibility level, we'll use a range within the assembly version. This gives us room for 1000 releases before we need to bump anything. Platform specific versions would use the range that matches its [.NET Platform Standard version](https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-platform-standard.md#mapping-the-net-platform-standard-to-platforms):
 - `netstandard1.0`: `3.0.0`-`3.0.999`
 - `netstandard1.1`: `3.0.1000`-`3.0.1999`
 - `netstandard1.3`: `3.0.3000`-`3.0.3999`
 - `net45`: `3.0.0`-`3.0.999`
 - `net46`: `3.0.3000`-`3.0.3999`
 - `win8`: `3.0.1000`-`3.0.1999`
 - `wp8`: `3.0.0`-`3.0.999`
 - `win81`: `3.0.2000`-`3.0.2999`
 - `wpa81`: `3.0.2000`-`3.0.2999`
 - `uap10.0`: `3.0.4000`-`3.0.4999`
 - `netcoreapp1.0`: `3.0.6000`-`3.0.6999` 

Given this model, we would increment `Minor` and `Patch` versions on the NuGet package as appropriate for the release, but each release would increment a value in the given range for the assembly. This will ensure that something isn't accidentally redirected to a higher assembly version with fewer/missing API surface area.

One open question is `Major` versions. Given that major versions are generally considered able to have breaking changes, it may be fair to bump the major assembly version to match. This could potentially lead to an issue if a `3.0.3000` assembly gets redirected to a `4.0.0` version that has less surface area. For a major release, this might be acceptable.

Thoughts? 

/cc @bartdesmet @paulcbetts @mattpodwysocki @shiftkey @LeeCampbell @shana @forki @OmerRaviv @natemcmaster

### PR DESCRIPTION
**Pull request by onovotny**
Fixes #205 and #97.

/cc @shiftkey @mattpodwysocki @natemcmaster @shana @forki @omerraviv for review.

---
This change is [Reviewable](https://reviewable.io/reviews/reactive-extensions/rx.net/212)

### CODE REVIEW

**Comment by shiftkey** [comment]
Trying to build the project errors out here. Am I missing something?

```
Attempting to gather dependency information for package 'gitversion.commandline.3.5.5-pullrequest0921' with respect to project 'packages', targeting 'Any,Version=v0.0'
Package 'gitversion.commandline,3.5.5-pullrequest0921' is not found in the following primary source(s): 'C:\Users\shiftkey\AppData\Local\NuGet\Cache,C:\Users\shiftkey\.nuget\packages\,https://api.nuget.org/v3/index.json,C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\'. Please verify all your online package sources are available (OR) package id, version are specified correctly.
```

**Comment by onovotny** [comment]
It should be using the Build Packages feed with the guid in it.... That's where I put that package while I wait for it to be officially merged and published.

**Comment by shiftkey** [comment]
@onovotny I think this is related to where you execute the `build-new` scripts from.

On my machine, when I run `.\Rx.NET\Source\build-new` the location isn't changed, so it:

 - checks and downloads nuget.exe to root of the repository
 - `nuget restore ...` doesn't resolve a `nuget.config` at the root
 - it misses the additional `Build Packages` feed, so it can't find this new package

The workaround I tested was to do this:

```powershell
Push-Location ".\Rx.NET\Source\"
Write-Host "Building Rx.NET" -ForegroundColor Green
.\build-new
Pop-Location
```

There's probably a much cleaner way to do this, but I also haven't looked into why Appveyor isn't failing in the same way...

**Comment by onovotny** [comment]
@shiftkey there's a nuget.config at the root here that should work https://github.com/Reactive-Extensions/Rx.NET/blob/master/NuGet.Config. Was that not getting picked up for you?

File: Ix.NET/Source/Ix.NET.sln [hunk 1]
```diff
@@ -17,7 +17,8 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "System.Interactive.Tests",
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution Items", "{B733D97A-F1ED-4FC3-BF8E-9AC47A89DE96}"
 	ProjectSection(SolutionItems) = preProject
-		build-new.ps1 = build-new.ps1
+		GlobalAssemblyVersion.cs = GlobalAssemblyVersion.cs
+        build-new.ps1 = build-new.ps1
```
Comment (shiftkey) [inline Ix.NET/Source/Ix.NET.sln]: :lipstick: indenting (lol)

File: Rx.NET/Source/System.Reactive.Observable.Aliases/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+﻿using System;
```
Comment (shiftkey) [inline Rx.NET/Source/System.Reactive.Observable.Aliases/Properties/AssemblyInfo.cs]: Is there a particular reason why we need this file?

**Comment by shiftkey** [comment]
> Was that not getting picked up for you?

It was not, but let me have another shot at it now...

File: Rx.NET/Source/System.Reactive.Observable.Aliases/Properties/AssemblyInfo.cs (hunk 2 above)
Comment (onovotny) [inline Rx.NET/Source/System.Reactive.Observable.Aliases/Properties/AssemblyInfo.cs]: this file was missing before actually...so this assembly never had its attributes set.

File: Ix.NET/Source/Ix.NET.sln (hunk 1 above)
Comment (onovotny) [inline Ix.NET/Source/Ix.NET.sln]: bah, I'll fix it -- was a merge conflict, so VS didn't generate that.

**Comment by shiftkey** [comment]
Build script is fine now on my machine, I just can't create packages:

```
Building Packages

    Directory: C:\Users\shiftkey\Documents\GitHub\rx.net


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----       11/07/2016   8:07 AM                artifacts
Attempting to build package from 'Microsoft.Reactive.Testing.nuspec'.
The special version part cannot exceed 20 characters.
Attempting to build package from 'System.Reactive.Core.nuspec'.
The special version part cannot exceed 20 characters.
Attempting to build package from 'System.Reactive.Experimental.nuspec'.
The special version part cannot exceed 20 characters.
Attempting to build package from 'System.Reactive.Interfaces.nuspec'.
The special version part cannot exceed 20 characters.
Attempting to build package from 'System.Reactive.Linq.nuspec'.
The special version part cannot exceed 20 characters.
Attempting to build package from 'System.Reactive.nuspec'.
The special version part cannot exceed 20 characters.
Attempting to build package from 'System.Reactive.Observable.Aliases.nuspec'.
The special version part cannot exceed 20 characters.
Attempting to build package from 'System.Reactive.PlatformServices.nuspec'.
The special version part cannot exceed 20 characters.
Attempting to build package from 'System.Reactive.Providers.nuspec'.
The special version part cannot exceed 20 characters.
Attempting to build package from 'System.Reactive.Runtime.Remoting.nuspec'.
The special version part cannot exceed 20 characters.
Attempting to build package from 'System.Reactive.Windows.Forms.nuspec'.
The special version part cannot exceed 20 characters.
Attempting to build package from 'System.Reactive.Windows.Threading.nuspec'.
The special version part cannot exceed 20 characters.
Attempting to build package from 'System.Reactive.WindowsRuntime.nuspec'.
The special version part cannot exceed 20 characters.
```

Looking at the diff for the generated `AssemblyInformationVersion` entry, it seems like GitVersion is being very verbose:

```
-[assembly: AssemblyFileVersion("2.2.0.0")]
-[assembly: AssemblyInformationalVersion("2.2.0.0")]
+[assembly: AssemblyFileVersion("3.0.1.0")]
+[assembly: AssemblyInformationalVersion("3.0.1-assembly-versions.27+Branch.assembly-versions.Sha.ab35480481315a6703909afab95b4fb769b6d898")]
```

Do we care about this constraint?

**Comment by shiftkey** [comment]
After a fresh build there are two changes to dependencies:

- `Rx.NET/Source/NuSpecs/Microsoft.Reactive.Testing.nuspec` bumps `xunit.assert` from v2.1.0 to 2.2.0-beta2-build3300
- `Rx.NET/Source/NuSpecs/System.Reactive.Observable.Aliases.nuspec` adds `System.Resources.ResourceManager` v4.0.1

**Comment by shiftkey** [comment]
@onovotny just a few little things - would love to be able to poke the packages and verify the changes but I won't hold that up because of how GitVersion names branches in this situation...

**Comment by onovotny** [comment]
Yeah, I wouldn't worry about any package failures due to branch names. I wish NuGet didn't have that limit; otherwise, I can also look into ways to truncate to fit. Prob another PR.

**Comment by onovotny** [comment]
As for versions of BCL refs, I'm going to create a separate PR to remove those granular deps and just reference NETStandard.Library. That's the new/current recommendation of the NuGet/BCL team.

**Comment by shiftkey** [comment]
@onovotny :thumbsup:

**Comment by onovotny** [comment]
As for the `xunit.assert` version, #210 will turn that into a source reference, internal, so that dependency will go too.

**Pull request merged**